### ISSUE TITLE
New Development Workflow Proposal

### ISSUE THREAD

**Issue by erik-whiting** [RNAMake/RNAMake#6]
For everyone involved in writing code for RNAMake (or any other project in the lab), I think we should decide on a standardized workflow. I will propose a workflow here, please feel free to make your own proposal, and ask questions or make comments on this one using the comment feature.

# Overview

The workflow I'm about to propose is based on a workflow called "trunk based development." Trunk based development is common in DevOps and Agile development methodologies and allows for strong synchronization among all members of a project team. I'll go into detail on how it works and then talk about how we could tweak the process for our purposes.

# Details

In this section, I'll explain trunk based development in detail, then explain a slight modification for our specific uses.

## Trunk Based Development

In trunk based development, there is one main branch upon which all development work is based. Usually, this branch is called `master` or `main`. Let's stick with `main` for the rest of this proposal.

From one developer's perspective, there are two copies of `main`: the remote one (the one that exists here on GitHub), and the local one (the one that exists on the developer's personal computer). When a developer is about to do work, they will first _checkout_ the `main` branch. This means they will synchronize their local `main` branch with the remote `main` branch.

Once the developer has the latest version of `main` on their local machine, the developer will then _branch off of_ `main`. This means that the developer will create a new branch, of which `main` will be the _base_. Let's call this new branch `add-new-feature`.

This means that the `add-new-feature` branch will use the `main` branch as something of a starting point. Any work done on the `add-new-feature` branch can be rolled back to the `main` branch at any time. If you're into video games, you can think of `main` as a save file that you can always fall back on if things go terribly wrong.

The developer does their work on the `add-new-feature` branch and then _commits_ that work to the branch. This means that the developer has essentially saved that work to the `add-new-feature` branch and that work is the new starting point of that branch. Some work will require lots of commits, some work will require just one commit. You can also think of commits as save files, points in time to which you can rollback in case you need to.

Once the developer is happy with their work, they will _push_ the `add-new-feature` branch to the repository. When a developer pushes a branch, a copy of their local branch (in this case, `add-new-feature`) is added to the remote repository (RNAMake on GitHub in this example).

Once the developer's branch has been pushed, they make a _pull request_. You can think of this as submitting an article for peer review, except instead of an article, it's code. Pull requests are sometimes called _merge requests_, so if you see that term used, just know it's the same thing.

A pull request is a developer's way of officially submitting their code and asking for it to be _merged_ into the `main` branch (more on merging later). This is when _code review_ starts.

Code review is the process of other developers on a team looking over the developer's pull request. After reviewing what's been added, deleted, or changed, other developers will ask questions for clarity, suggest different approaches, or look for mistakes in the new branch. After review, the developer will respond to the comments by either answering them on the pull request or making changes in the code, then resubmitting for review. This cycle repeats until the changes are approved.

Once everyone is happy with the changes proposed in the pull request, the project's maintainer merges the `add-new-feature` branch into the `main` branch. Merging is how new changes on a branch are added to the `main` branch. Once merged into `main`, all the other developers on that team have access to those changes and simply need to pull the remote `main` branch to their local `main` branch. This way, the next time someone starts doing new work off of the `main` branch, the changes from `add-new-feature` will be there.

This is probably an unnecessarily detailed description of trunk based development, once you've done it a few times, you will get the hang of it, I just wanted to make sure I explained every detail to avoid as much confusion as possible. If you have any questions, please post them as comments here or feel free to send me a private message in Slack.

## Tweaking Trunk Based Development

I have noticed that there is a `master` branch (which would be the trunk or the equivalent of what I called the `main` branch in the previous section), and several `abc/structure_rework` branches. I assume this is because the `master` branch represents a version release that is currently consumed by the public.

Having the `master` branch be the only one "exposed" to the public is normal, but having multiple branches that are likely out-of-sync with each other is prone to merge conflicts and many other problems. Instead, I propose the following tweak:

We make a new branch based off of `master` called `structure-rework` or something similar. Then, we treat that branch like the `main` branch in the example in the previous section. Every time someone wants to add a single feature or fix a single bug, they make a branch specifically for that work, and that work only. Once that work is done, reviewed, and approved, we merge it into the `structure-rework` branch. As the rework goes on and everyone branches off of the shared `structure-rework` branch, we rarely (maybe never) have work that is out of sync with other developers in the lab (please trust me, merging long running unsynchronized branches together is an absolute headache, best to avoid it).

Once the rework is completed, we can merge that branch into `master` and perhaps create a new release to be consumed by the public. Doing this will help ensure that none of the structure rework we do creates merge conflicts with the `master` branch and publishing a new release is a relatively straightforward task.

# Benefits

There are a few benefits to using the trunk based development workflow, which I'll briefly go over here.

**No Merge Conflicts:** If we follow the workflow I've described, everyone will be able to work off of the most recent changes at all times. This means there will be no merge conflicts once the rework is done, saving us lots of work when it's time to make a new release.

**Better Collaboration:** If we're merging small branches to the main branch as the work is ready, every developer always has access to the latest available features. This means everyone is building together rather than in different places.

**Industry Standard:** Trunk based development is the most common way companies do software development. If anyone working on this project ever finds themselves working in a software-intensive environment, you will be glad you already know how to use git and what the typical workflow for getting code to customers is.

# Conclusion

That concludes my proposal for how we should do development from now on. Please feel free to leave feedback or ask questions in the comments. You can also DM me in Slack or send me an email if you want to ask/say something to just me. Thank you!
